tool/tctl/common/resource_command.go

```diff
@@ -517,12 +518,34 @@ func (rc *ResourceCommand) Delete(client auth.ClientI) (err error) {
 			return trace.Wrap(err)
 		}
 		fmt.Printf("kubernetes service %v has been deleted\n", rc.ref.Name)
+	case services.KindClusterAuthPreference:
+		if err = resetAuthPreference(ctx, client); err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Printf("cluster auth preference has been reset to defaults\n")
 	default:
 		return trace.BadParameter("deleting resources of type %q is not supported", rc.ref.Kind)
 	}
 	return nil
 }
 
+func resetAuthPreference(ctx context.Context, client auth.ClientI) error {
+	storedAuthPref, err := client.GetAuthPreference()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	managedByStatic := storedAuthPref.Origin() == types.OriginConfigFile
+	if managedByStatic {
+		return trace.BadParameter(managedByStaticDeleteMsg)
+	}
+
+	if err = client.ResetAuthPreference(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
```

Comment:
Wouldn't it be more appropriate to keep the two rather similar `managedByStatic` messages together? My idea with the client-side validation was to be able to provide a more detailed, potentially `tctl`-specific message.

Reply:
Fair point.
Can `ResetAuthPreference` return a `FailedPrecondition` (or some other distinct) gRPC status code when this validation fails?
That way you could print the matching message and also avoid duplicating validation logic.

Reply:
Logically the validation belongs to `auth.ServerWithRoles`, not to the gRPC handler.

There is analogy between how the case of `origin: config-file` is handled in `SetAuthPreference` (by requiring a privilege) and how it's handled in `ResetAuthPreference` (that altogether rejects reset attempts in this case).

On the client side, the analogy would imply the auth preference should be pre-fetched to display the best possible user message, because that's how it's done with `SetAuthPreference`.

Reply:
OK, I guess it's not very clear whether checking `origin` is _validation_ (belongs in `auth.Server`) or _authorization_ (belongs in `auth.ServerWithRoles`).
It definitely does not belong in the gRPC handler (`auth.GRPCServer`). Let's keep it in `ServerWithRoles` for now.